### ISSUE TITLE
Modal open property only updates once after update to 4.6.0

### ISSUE THREAD

**Issue by kboucher** [kaliber5/ember-bootstrap#1386]
I recently updated from `ember-bootstrap@4.4.0` to `ember-bootstrap@4.6.0` in order to pick up the fix for [issue 1362](https://github.com/kaliber5/ember-bootstrap/issues/1362).

I notice that in addition to that fix, the updates include converting the Modal component to a Glimmer component.

In any case, it seems that the `@open` property is now only updating one time. I have an event handler for the `onHidden` event that toggles the property I am passing to `@open`, however, after opening and closing the modal one time, the modal no longer displays even though I can see the toggle property and the `@open` property in the `BsModalComponent` updating as expected in the Ember Inspector.

This is also my first Octane project, so I'm not entirely confident that I haven't done something wrong, but my setup worked in 4.4.0, and I have gone back to that version for now.

I will try to return to this ticket with any more useful information I can discern but wanted to raise the issue now.

**UPDATE**

I ran a `git bisect` and have identified the following commit as the _first bad commit_:

```
1bf4cf7aa2fe9aefa4e618b601877e1d4b0b5843 is the first bad commit
commit 1bf4cf7aa2fe9aefa4e618b601877e1d4b0b5843
Author: simonihmig <simon.ihmig@kaliber5.de>
Date:   Sun Dec 20 22:00:33 2020 +0100

    Call handleVisibilityChanges on public @open change

 addon/components/bs-modal.hbs | 2 +-
 addon/components/bs-modal.js  | 6 +++---
 2 files changed, 4 insertions(+), 4 deletions(-)
```

It also seems to be the case that when the modal fails to open on subsequent updates to the `@open` property, the following error is observed:

```
Uncaught Error: Assertion Failed: Backdrop element should be in DOM
```

Additionally, I've noticed that there is some undesired behavior related to the backdrop. (The commit where this arises is: `56ab45ed7f3ec1f01993421fd227d1feacbd5319`)

Specifically, the `.show` class is removed from the backdrop before the modal is de-rendered -- causing a solid black backdrop for a few seconds until the modal is de-rendered.

My implementation is pretty straight-forward:

`NavigationDrawerComponent`:
```
<BsModal
    class="navigation-drawer {{@side}}"
    @fade={{false}}
    @onHidden={{@onNavigationDrawerHidden}}
    @onShown={{this.onNavigationDrawerShown}}
    @open={{@showNavigationDrawer}}
    @scrollable={{true}}
    as |Modal|
>
    <Modal.body>
        {{yield}}
    </Modal.body>
</BsModal>
```

`@onNavigationDrawerHidden` is passed in from the parent component (`MainNavigationComponent`) and references a service (`MainNavigationService`) that toggles the property `showMainMenu` which is also passed in from the `MainNavigationComponent`: 

```
<NavigationDrawer
    @onNavigationDrawerHidden={{this.mainNavigation.toggleMainMenu}}
    @showNavigationDrawer={{this.mainNavigation.showMainMenu}}
    @side="left"
> ... </NavigationDrawer>
```

**Comment by lindyhopchris** [comment]
I'm seeing this problem as well. Specifically, when the modal gets closed, this error occurs:

```
index.js:172 Uncaught Error: Assertion Failed: Backdrop element should be in DOM
    at Object.assert (index.js:172)
    at wrapperClass.handleBackdrop (bs-modal.js:484)
    at wrapperClass.hideModal (bs-modal.js:445)
    at bs-modal.js:427
    at invokeCallback (rsvp.js:493)
    at publish (rsvp.js:476)
    at rsvp.js:19
    at invoke (backburner.js:338)
    at Queue.flush (backburner.js:229)
    at DeferredActionQueues.flush (backburner.js:426)
```

The reason why the modal can then not be opened a second time is because this error occurs *before* the `onHidden` callback is triggered. So because of the error, the `@opens` property is not toggled.

I was also on EB 4.4 and encounter this when upgrading to 4.6

**Comment by jelhan** [comment]
I'm able to reproduce this issue in a new Ember application created with Ember CLI 3.24 using latest release of Ember Bootstrap. I used [ember-set-helper](https://github.com/pzuraq/ember-set-helper) and this template to test:

```hbs
{{#let (hash showModal=false) as |state|}}
  <button {{on "click" (set state "showModal" true)}}>show modal</button>

  <BsModalSimple
    @open={{state.showModal}}
    @onHidden={{set state "showModal" false}}
  >
    My Modal
  </BsModalSimple>
{{/let}}
```

I see the error included in the bug report from @lindyhopchris regardless if the modal is closed using the button or by a click on backdrop.

Their are two assertions with the same text. In both cases this assertion fails: https://github.com/kaliber5/ember-bootstrap/blob/acaf5ec4fe00713eda37025304139024ed843c10/addon/components/bs-modal.js#L534-L536

The function is executed twice for the same click on the close button. `this.backdropElement` is a `HTMLElement` for the first pass but not on the second. I _guess_ the main issue is that the function should not be called twice.

This reminded me of a bug, which was fixed some months ago: `@onDismissed` being called twice for `<BsAlert>` (#1264). I _guess_ we would see `@onHidden` being called twice as well if would not throw before.

I _guess_ `this.handleVisibilityChanges()` is executed when setting the value passed as `@open` to `false` in event handler bound to `@onHidden`: https://github.com/kaliber5/ember-bootstrap/blob/acaf5ec4fe00713eda37025304139024ed843c10/addon/components/bs-modal.hbs#L2 https://github.com/kaliber5/ember-bootstrap/blob/acaf5ec4fe00713eda37025304139024ed843c10/addon/components/bs-modal.js#L681-L693

I guess we need to guard add a guard against this in `handleVisibilityChanges` or `hide` methods of `Modal` class similar as how it was done in the fix for `Alert`. See #1281

**Comment by jelhan** [comment]
The `{{did-update}}` helper seems to fire without `@open` being changed. Would love to share an Ember Twiddle as it very difficult to explain. But if you put a debugger inside of `handleVisibilityChanges`, you will notice that it's executed immediately after clicking the close button of a modal. If you put an additional debugger into the `{{did-update}}` helper, you will notice that it's causing that execution.

That sounds like an issue with auto-tracking causing invalidation more often than one would expect. One of the reasons why this observer style pattern using `ember-render-helpers` or `ember-render-modifiers` is problematic. But it is nevertheless very confusing that `@open` argument is invalidated without any change to the value bound to it. :confused: 

My first guess is that the issue is related to `localCopy` decorator from `tracked-toolbox`, which is used with `@open` argument. https://github.com/kaliber5/ember-bootstrap/blob/acaf5ec4fe00713eda37025304139024ed843c10/addon/components/bs-modal.js#L79-L80 The local `isOpen` state is set immediately before the `{{did-update}}` helper is fired.

**Comment by simonihmig** [comment]
Fixed by #1393
